vector/src/main/java/im/vector/riotx/features/login/LoginActivity.kt

```diff
@@ -18,64 +18,202 @@ package im.vector.riotx.features.login
+        loginLoading.isVisible = loginViewState.isLoading()
+    }
+
+    private fun onWebLoginError(onWebLoginError: LoginNavigation.OnWebLoginError) {
+        // Pop the backstack
+        supportFragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
+
+        // And inform the user
+        AlertDialog.Builder(this)
+                .setTitle(R.string.dialog_title_error)
+                .setMessage(getString(R.string.login_sso_error_message, onWebLoginError.description, onWebLoginError.errorCode))
+                .setPositiveButton(R.string.ok, null)
+                .show()
+    }
+
+    private fun onServerSelectionDone() {
+        when (loginViewModel.serverType) {
+            ServerType.MatrixOrg -> Unit // In this case, we wait for the login flow
+            ServerType.Modular,
+            ServerType.Other     -> addFragmentToBackstack(R.id.loginFragmentContainer, LoginServerUrlFormFragment::class.java)
+        }
+    }
+
+    private fun onSignModeSelected() {
+        when (loginViewModel.signMode) {
+            SignMode.Unknown -> error("Sign mode has to be set before calling this method")
+            SignMode.SignUp  -> addFragmentToBackstack(R.id.loginFragmentContainer, LoginFragment::class.java)
+            SignMode.SignIn  -> {
+                // It depends on the LoginMode
+                withState(loginViewModel) {
+                    when (val loginMode = it.asyncHomeServerLoginFlowRequest.invoke()) {
+                        null                     -> error("Developer error")
+                        LoginMode.Password       -> addFragmentToBackstack(R.id.loginFragmentContainer, LoginFragment::class.java)
+                        LoginMode.Sso            -> addFragmentToBackstack(R.id.loginFragmentContainer, LoginWebFragment::class.java)
+                        is LoginMode.Unsupported -> onLoginModeNotSupported(loginMode)
+                    }
+                }
             }
         }
+    }
+
+    private fun onLoginModeNotSupported(unsupportedLoginMode: LoginMode.Unsupported) {
+        AlertDialog.Builder(this)
+                .setTitle(R.string.app_name)
+                .setMessage(getString(R.string.login_mode_not_supported, unsupportedLoginMode.types.joinToString { "'$it'" }))
+                .setPositiveButton(R.string.yes) { _, _ ->
+                    addFragmentToBackstack(R.id.loginFragmentContainer, LoginWebFragment::class.java)
+                }
+                .setNegativeButton(R.string.no, null)
+                .show()
+    }
 
-        loginViewModel.selectSubscribe(this, LoginViewState::asyncLoginAction) {
-            if (it is Success) {
-                val intent = HomeActivity.newIntent(this)
-                startActivity(intent)
-                finish()
+    private fun handleRegistrationNavigation(flowResult: FlowResult) {
+        // Complete all mandatory stage first
+        val mandatoryStages = flowResult.missingStages.filter { it.mandatory }
```

Comment:
Looks like you only need `firstOrNull` here.

Reply:
Yes, indeed :)